crates/core/src/sync/streaming_sync.rs

```diff
@@ -448,6 +448,12 @@ impl StreamingSyncIteration {
                     |s| s.start_tracking_checkpoint(progress, subscription_state),
                     &mut event.instructions,
                 );
+
+                // Technically, we could still try to apply a pending checkpoint after receiving a
+                // new one. However, sync_local assumes it's only called in a state where there's no
+                // pending checkpoint, so we'd have to take the oplog state at the time we've
+                // originally received the validated-but-not-applied checkpoint. This is likely not
+                // something worth doing.
```

Comment:
> This is likely not something worth doing.

The current implementation effectively "invalidates" any past checkpoint whenever we receive new data - we do not keep the data around to apply a past checkpoint. While we could implement something to keep the last checkpoint around, it is a big project, so I agree with "likely not worth doing" at least for now.